### PR TITLE
Gitlab backend

### PR DESCRIPTION
**Pull request by jonathangold**
This PR adds a gitlab backend. Still work in progress, will ping someone for review when it's ready.

### CODE REVIEW

**Comment by jonathangold** [comment]
@knelasevero The docker stage is failing at

```
#42 [linux/arm/v7 stage-1 1/4] FROM gcr.io/distroless/base-debian10:latest
#42 resolve gcr.io/distroless/base-debian10:latest 0.1s done
#42 ERROR: no match for platform in manifest: not found
```

https://github.com/ContainerSolutions/externalsecret-operator/pull/131/checks?check_run_id=1370440971#step:7:86

File: pkg/gitlab/backend.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
+// Package dummy implements an example backend that can be used for testing
```
Comment (jonatasbaldin) [inline pkg/gitlab/backend.go:1]: Awesome mate! I think you could remove this comment tho, related to the `dummy` backend.

File: README.md [hunk 2]
```diff
@@ -337,6 +344,90 @@ spec:
...
+```
+
+<a name="gitlab-cicd-variables"></a>
+
+## Gitlab CI/CD Variables
+
+<a name="gitlab-cicd-variables-pre"></a>
+
+#### Prerequisites
+- A Gitlab project with a CI/CD variable, who's key is `example-externalsecret-key`
```
Comment (jonatasbaldin) [inline README.md:361]: Hey mate, I got an error while creating a variable with this name: `Variables key can contain only letters, digits and ‘_’.`. 

What about changing the example to `example_externalsecret_key`, with underscores?

Screenshot 2020-11-12 at 11 47 50

File: README.md [hunk 3]
```diff
@@ -337,6 +344,90 @@ spec:
...
+- A [Gitlab personal access token](https://gitlab.com/-/profile/personal_access_tokens)
```
Comment (jonatasbaldin) [inline README.md:362]: When creating a Gitlab Personal Access Token, I’m presented with Scope options for the key, such as `api`, `read_user` etc.

Could you document which one should be enabled, preferably the less permissive one.

File: README.md [hunk 4]
```diff
@@ -337,6 +344,90 @@ spec:
...
+spec:
+  controller: staging
+  store:
+    type: gitlab
+    auth:
+      secretRef:
+        name: externalsecret-operator-credentials-gitlab
+    parameters:
+      baseURL: https://gitlab.com
+      projectID: 12345678
```
Comment (jonatasbaldin) [inline README.md:411]: Could you document that the user also needs the Project ID and where to find it, on the Prerequisites section.

Thanks!

File: README.md [hunk 5]
```diff
@@ -337,6 +344,90 @@ spec:
...
+    name: externalsecret-operator-secretstore-sample
+  data:
+    - key: example-externalsecret-key
+      version: latest
+```
+
+- The operator fetches the CI/CD variable from Gitlab and injects it as a secret:
+
 ```shell
 % make install
```
Comment (jonatasbaldin) [inline README.md:432]: Do we need to run `make install` again? It’s already the first step of the guide. If not, we can remove here or remove the first one on top.

What do you think?

File: pkg/gitlab/backend.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+	d.client, err = gitlab.NewClient(gitlabCreds.Token, gitlab.WithBaseURL(baseURL.(string)))
+	if err != nil {
+		log.Error(fmt.Errorf("error"), "failed to create client: ")
+		return fmt.Errorf("failed to create client")
+	}
+	return nil
+}
+
+// Get takes a key and version, and returns the value
+func (d *Backend) Get(key string, version string) (string, error) {
```
Comment (jonatasbaldin) [inline pkg/gitlab/backend.go:86]: Should we also `log` if the `Get` operation was successful? What do you think?

**Review by jonatasbaldin — CHANGES_REQUESTED** [review]
Hey mate, thanks a lot for this work! 

I tested and it worked just fine ✨ Also left some comments, mostly about the guide in the README section.

File: README.md (hunk 2 above)
Comment (jonathangold) [reply README.md:361]: Oh, I forgot gitlab doesn't support dashes. (I tested with the name test.) Will change

File: README.md (hunk 3 above)
Comment (jonathangold) [reply README.md:362]: will do. I think API is enough

File: README.md (hunk 5 above)
Comment (jonathangold) [reply README.md:432]: It's listed as a prereq in the gsm section as well. If we remove it here, we should remove it there as well. What do you think?

File: pkg/gitlab/backend.go (hunk 6 above)
Comment (jonathangold) [reply pkg/gitlab/backend.go:86]: Not a bad idea :)

File: README.md (hunk 5 above)
Comment (jonatasbaldin) [reply README.md:432]: If it's not required, I think it's ok to remove everywhere.

**Comment by jonatasbaldin** [comment]
Thanks for all the replies @jonathangold!

We just merged the removal of the 1Password backend and legacy code... could u rebase this branch too? ✨

Thanks ❤️

File: README.md (hunk 3 above)
Comment (jonathangold) [reply README.md]: I put `read_api` in the doc, but I still have to verify that it's enough.

**Comment by jonathangold** [comment]
@jonatasbaldin Thanks for the review! I made the changes you pointed out. Just need to validate the api permissions are enough.

File: README.md (hunk 3 above)
Comment (jonatasbaldin) [reply README.md]: I think that should be it too!

Comment (jonatasbaldin) [reply README.md]: Let me know when u change it so we can merge this code :)

File: pkg/gitlab/backend.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	if key == "" {
+		return "", fmt.Errorf("empty key provided")
+	}
+
+	variable, _, err := d.client.ProjectVariables.GetVariable(fmt.Sprintf("%.f", d.projectID), key, nil)
+	if err != nil {
+		return "", err
+	}
+
+	fmt.Printf("Get was successful for the Gitlab variable with key: %s\n", key)
```
[the remark this answers is not in the document]
Comment (jonathangold) [reply pkg/gitlab/backend.go]: Tried to use `log.Printf()`, but realized log is an instance of the k8s controller logger, not `"log"`, so I probably need to use log.Info, but I'm away from a working IDE at the moment. :/ will rebase and fix soon.

[the remark this answers is not in the document]
Comment (jonatasbaldin) [reply pkg/gitlab/backend.go]: oki :)

[the remark this answers is not in the document]
Comment (jonathangold) [reply pkg/gitlab/backend.go]: @jonatasbaldin it's fixed.

File: README.md (hunk 3 above)
Comment (jonathangold) [reply README.md]: @jonatasbaldin read_api works. :+1:

**Pull request merged**